### PR TITLE
Fix ROOT-10484 by implementing dyld support for RPATH.

### PR DESCRIPTION
**Pull request by vgvassilev**
Some libraries are layered can depend on other libraries on a private paths. That is, libA can depend on libB which is neither on the LD_LIBRARY_PATH nor on a known system path. The posix linker injects "variables" such as @rpath which is expanded at link time to resolve the libraries on a relative path.

Prior to this patch, cling's Dyld-based symbol resolution could not trace down such cases causing failures in symbol resolution when a symbol is only defined in libB (a private library).

This patch implements the basic posix linker substitutions allowing cling's Dyld-based symbol resolution implementation to follow more closely the linker rules.

## Performance

|platform|test|master|PRv1|PRv2| mem master|mem PRv2|
|:--------|:---|:---------:|:-----------:|:-----------:|:-----------:|:-----------:|
|ubu20|roottest-python-pythonizations|6.5|6.55|6.65|342076|347168|
|ubu20|roottest-cling|112.8|114.53|181.26|411048|416036|
|ubu20|roottest-python|148.5|136.26|144.72|394292|397120|
|ubu20|roottest-root-hist|16.78|17.35|15.45|465156|235084|
|ubu20|roottest-root-math|28.28|27.89|27.40|424072|371256|
|ubu20|roottest-root-tree|462.84|510.59|465.79|796512|797992|
|ubu20|roottest-root-treeformula|117.53|129.16|114.44|261184|260168|
|ubu20|root-io-stdarray|66.90|67.89|65.71|263424|255120|
|ubu20|roottest-root-treeformula|117.55|129.10|
|osx 10.15|roottest-python-pythonizations|17.25|16.06|
|osx 10.15|roottest-cling|346.06|343.46|
|osx 10.15|roottest-root-hist|36.06|36.22|
|osx 10.15|roottest-root-math|49.78|52.42|
|osx 10.15|root-io-stdarray|119.32|118.73|

That PR should reduce the number of `stat` calls improving #7774

Kudos Alexander Penev (@alexander-penev).

### CODE REVIEW

**Comment by vgvassilev** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Review by oshadura — COMMENTED** [review]
@vgvassilev wow! changes are so minimal, we always have this feature and never tested it? I afraid I am not sure if CI has R enabled (I see `-Dr=Off` is everywhere)

**Comment by vgvassilev** [comment]
The intent was that we only wanted to see what was failing. We are working on a bigger patch which will come in this PR.

**Comment by vgvassilev** [comment]
@phsft-bot build!

**Comment by vgvassilev** [comment]
@phsft-bot build!

File: interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h
Comment (Axel-Naumann) [inline interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h:91]: What does "library specific" mean? Is this maybe:
```suggestion
    ///\param[in] RPATH - RPATH as provided by another library, searching for libStem
```
? (Same comment for all other occurrences of "library specific")

File: interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp:196]: Pre-existing spell-o:
```suggestion
      SplitPaths(RPATH, RPaths, utils::kPruneNonExistent, platform::kEnvDelim,
```

File: interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp:202]: Shouldn't we at least replace `$loader_path` with `mainExecutable`, possibly keeping the (possibly modified) FIXME if still needed?

File: interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp:200]: IIUC `@executable_path` expects the *directory* of the binary location, while `getMainExecutable()` returns the full path to the image file. I.e. there's "get the directory of `mainExecutable`" missing somewhere?

File: interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp:163]: Maybe
```suggestion
  static bool substFront(llvm::StringRef value, std::string find,
```
It's a key property of this function that it only replaces text at the *beginning* of the string.

Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp:163]:
```suggestion
  static bool subst(llvm::StringRef original, llvm::StringRef pattern,
```

File: interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp:164]:
```suggestion
                    llvm::StringRef replacement, std::string& result) {
```

Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp:164]: The return value of this function isn't ever used. Consider
```c++
static std::string substFront(llvm::StringRef value, llvm::StringRef pattern, llvm::StringRef replacement);
```

File: interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp:165]: If you decide to return `string`, consider inverting:
```suggestion
    if (!value.startswith_lower(find))
      return find;
```

File: interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp:167]: Why? If I read this correctly, replacing "foo/" with "baz/" in "foo/bar" will result in "baz//bar"?

File: interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h
Comment (Axel-Naumann) [inline interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h:84]: Should these be `StringRef`s?

File: interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp:89]: Why is the order of priority

1. RPATH
2. user-provided paths and LD_LIBRARY_PATH
3. RUNPATH?

Maybe worth a comment?

File: interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h
Comment (Axel-Naumann) [inline interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h:84]:
```suggestion
                                 const std::string& RPath = "",
                                 const std::string& RunPath = "") const;
```

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:439]:
```suggestion
    if (!P2 || !P2[0])
```

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:529]: Would it make sense to register the library *with* its dependencies, instead of registering it, finding it back, const-casting what we found, and then providing the dependencies?

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:587]: For experiments, with ten thousand libraries, and path lengths of 1k, and in average 20 dependent libraries each: how much memory are we talking about here?

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:510]:
```suggestion
        auto HandleLib =
```

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:559]: We shouldn't be looking at non-process ABI: if this process is 64bit, only look at 64bit ELF; if it's 32 only look at 32bit ELF. Same for endianness. Or what am I missing?

**Review by Axel-Naumann — CHANGES_REQUESTED** [review]
That's quite some code - much of that would be much better off being hosted by llvm :-/ Is it worth trying to move the RPath handling upstream?

The main issue I have with this is the lack of tests: for that much, fundamentally important code it's a real pity that we don't have any tests. Could those be added, please? (Even something ROOT-only would make me happy.) Thanks!

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:510]: `auto` only works for non-recursive lambda functions -- we need to fix the return type.

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:559]: We already seem to have that filtered by https://github.com/vgvassilev/cling/blob/f3768a4c43b0f3b23eccc6075fa178861a002a10/lib/Interpreter/DynamicLibraryManagerSymbol.cpp#L730-L731

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (Axel-Naumann) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:510]: Can you use trailing return instead, in the definition of the labmda? `auto` is just nicer (and FWIW actually more performant that having to wrap this in a `std::function`)

Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:510]: Nope: https://godbolt.org/z/3hxGs6K9v

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:577]: What about COFF? `DLL`s also have that info.

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (Axel-Naumann) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:527]: Oh, *recursive* - missed that (despite you calling it out), thanks!

File: interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h
Comment (vgvassilev) [reply interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h]: Should now be improved.

File: interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp]: Thanks fixed!

File: interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp]: Yeah, good point. Fixed.

Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp]: Resolved.

File: interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp]: Yes, that way is clearer.

Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp]: Yeah, good point. Fixed.

File: interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp]: We have now a comment with a link to the description of the priorities for the static linker.

File: interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h
Comment (vgvassilev) [reply interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h]: Fixed.

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:555]: That is a good point, however, we should recursively, in-depth (depth-first search) accumulate the dependency list. The current algorithm is consistent to the behavior of the static linker. Moreover, the recursion in the addition of the dependencies is at scanning time, which makes the search more efficient as we do not need to recurse.

The proposed change is likely implementable but will require reorganization and performance tuning -- a task that we are not sure it is worth to avoid a minor (in our point) issue.

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp]: See previous comment.

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:605]: Can you point us to where you found that information? We do not seem to have access to `rpath` & co in the `COFFObjectFile`.

Comment (Axel-Naumann) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:605]: With "that info" I meant the library dependencies which you collect for ELF as `DT_NEEDED`. IIUC an example for COFF is [here](https://github.com/llvm/llvm-project/blob/1128311a19179ceca799ff0fbc4dd206ab56e560/llvm/tools/llvm-objdump/COFFDump.cpp#L350)

File: interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h
Comment (Axel-Naumann) [reply interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h]: Great thanks!

File: interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp
Comment (Axel-Naumann) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp]: Yes please :-)

File: interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp:99]: Isn't this equivalent and less CPU and simpler logic?
```suggestion
    llvm::SmallString<512> loaderPath;
    if (libLoader.empty()) {
      loaderPath = llvm::sys::fs::getMainExecutable(nullptr, nullptr);
    } else {
      loaderPath = libLoader.str();
    }
    llvm::sys::path::remove_filename(loaderPath);
```

File: interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp:64]: Why not
```suggestion
  static std::string substFront(llvm::StringRef original, llvm::StringRef pattern,
```
? Looks like https://github.com/root-project/root/pull/6969#discussion_r618397300 wasn't applied properly, or I'm missing rationale.

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:614]: Why do you only `push_back` `nullptr`s?  Is the `if` condition meant to be inverted? This doesn't seem to be covered by a test?

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (Axel-Naumann) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp]: Could you please check with Shahzad for a profile before / after this PR is merged, both in *realtime* (due to many file ops) and *memory*?

**Review by Axel-Naumann — CHANGES_REQUESTED** [review]
Thank you for the improvements! Apart from the new comments here, please address https://github.com/root-project/root/pull/6969#discussion_r618406433

I'd still appreciate hearing your opinion on
> That's quite some code - much of that would be much better off being hosted by llvm :-/ Is it worth trying to move the RPath handling upstream?

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp]: @davidlange6, could you give us a hand?

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:605]: This pull request fixes a particular issue which in essence is a deployment technique on the unix platforms. I am hesitant adding such logic for COFF as I am not sure if that is relevant to that problem but also I am not sure what is the behavior of their dynamic linker. I'd pursue the approach from the past -- we will add the windows support if we need it in future.

**Comment by vgvassilev** [comment]
> Thank you for the improvements! Apart from the new comments here, please address [#6969 (comment)](https://github.com/root-project/root/pull/6969#discussion_r618406433)
> 
> I'd still appreciate hearing your opinion on
> 
> > That's quite some code - much of that would be much better off being hosted by llvm :-/ Is it worth trying to move the RPath handling upstream?
> 
> Isn't there existing "load that library" code anywhere in llvm, that either already handles RPATH or would be better off having RPATH support?

Yes, that is further down the road, the idea for this is to become a high-performance, default`llvm::orc::DefinitionGenerator`.

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (Axel-Naumann) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:605]: We *will* need the `DT_NEEDED` replacement for COFF to make ACLiC work when loading ACLiC libraries, without searching for libraries that can resolve unresolved symbols. I'm okay if this is addressed in a subsequent PR, but without that we cannot progress with https://github.com/root-project/root/issues/7774 - https://github.com/root-project/root/pull/5235 was excluding this case knowing that `DT_NEEDED` support would come.

File: interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManager.cpp:101]: We use the `mainExecutablePath` below.

**Review by Axel-Naumann — APPROVED** [review]
LGTM, thanks!

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:587]: The stats from cmssw look good.

File: interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp:577]: Let's do that, as we agreed, in another PR if required.

**Review by Axel-Naumann — CHANGES_REQUESTED** [review]
Please put Max back in relnotes

File: interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h
Comment (Axel-Naumann) [inline interpreter/cling/include/cling/Interpreter/DynamicLibraryManager.h:21]: Any chance that this can be
- removed, or
- moved into some namespace,
- and / or moved into the source?

**Comment by Axel-Naumann** [comment]
Failure in 32bit Linux roottest_python_cpp_cpp is unrelated

**Comment by vgvassilev** [comment]
@phsft-bot build!

**Comment by vgvassilev** [comment]
@phsft-bot build!

**Review by Axel-Naumann — APPROVED** [review]
LGTM once the warning on Ubuntu 16.04 is addressed!

**Pull request merged**
